File: qqkj_library/src/main/java/qqkj/qqkj_library/step/StepUtil.java
package qqkj.qqkj_library.step;

import android.content.Context;
import android.content.Intent;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;

/**
 * Created by apple on 2017/8/17.
 */

public class StepUtil {

    private Context _context;

    private static StepUtil _step_util;

    private Intent _intent = new Intent("_step_manager");

    /**
     * 构造函数,传入上下文对象
     * @param _context
     */
    private StepUtil(Context _context ){

        this._context = _context;

    }

    /**
     * 单例
     * @param _context
     */
    public static StepUtil getIns(Context _context ){

        if( _step_util==null ){

            _step_util = new StepUtil(_context);
        }

        return _step_util;
    }


    /**
     * 计步器方法
     *
     * 返回null,表示用户没有该传感器,返回_go_step,可以开始进行计步
     * 增加广播(名称:_step_manager)进行监听返回字段_step的值
     */
    public String _get_google_step(){

        SensorManager _manager = (SensorManager) _context.getSystemService(_context.SENSOR_SERVICE);

        Sensor _senor = _manager.getDefaultSensor(Sensor.TYPE_STEP_COUNTER);

        if(null==_senor){
            return "_step_fail";
        }

        SensorEventListener _sensorListener = new SensorEventListener() {
            @Override
            public void onSensorChanged(SensorEvent sensorEvent) {

                String _step = sensorEvent.values[0]+"";
                _intent.putExtra("_step",_step);
                _context.sendBroadcast(_intent);
            }

            @Override
            public void onAccuracyChanged(Sensor sensor, int i) {

                System.out.println("sensor = [" + sensor + "], i = [" + i + "]");
            }
        };

        _manager.registerListener(_sensorListener,_senor,Sensor.TYPE_STEP_COUNTER,SensorManager.SENSOR_DELAY_UI);

        return "_step_success";
    }
}

File: qqkj_library/src/main/java/qqkj/qqkj_library/sensor/step/StepUtil.kt
package qqkj.qqkj_library.sensor.step

import android.content.Context
import android.content.Intent
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager

/**
 * Created by chen2gou on 2017/8/24.
 *
 * 操作计步器方法
 */

class StepUtil {


    /**
     * 变量声明
     */
    var _context: Context? = null

    var _sensor_manager: SensorManager? = null

    var _sensor_listener: SensorEventListener? = null

    var _intent = Intent("_step_manager")

    var _sensor: Sensor? = null


    /**
     * 次构造函数声明
     */
    constructor(_context: Context?) {

        this._context = _context
    }


    /**
     * 静态块声明
     */
    companion object {

        var _step_util: StepUtil? = null

        fun getIns(_context: Context?): StepUtil {

            if (null == _step_util) {

                _step_util = StepUtil(_context)

            }

            return _step_util!!
        }
    }


    /**
     * 获取计步器步数方法
     */
    fun _get_step(): Boolean {

        _sensor_manager = _context!!.getSystemService(Context.SENSOR_SERVICE) as SensorManager

        _sensor = _sensor_manager!!.getDefaultSensor(Sensor.TYPE_STEP_COUNTER)

        if (null == _sensor) {

            //当前设备不存在计步传感器
            return false
        }

        _sensor_listener = object : SensorEventListener {

            override fun onSensorChanged(p0: SensorEvent?) {

                //获取计步器返回的运动步数
                var _step = p0!!.values[0]

                //将获取到的步数广播出去
                _intent.putExtra("_step", _step)

                _context!!.sendBroadcast(_intent)
            }

            override fun onAccuracyChanged(p0: Sensor?, p1: Int) {}
        }

        return true
    }


    /**
     * 回收计步传感器方法
     */
    fun _destroy_step() {

        if (null != _sensor_manager && null != _sensor_listener) {

            //取消注册计步器方法
            _sensor_manager!!.unregisterListener(_sensor_listener)
        }
    }
}